# 02 Data Understanding

---

> **Notebook:** `02_Data_Understanding.ipynb`  
> **Dataset:** Kaggle — Credit Card Fraud Detection  

---

## Objective

The objective of this notebook is to understand the structure and quality of the Credit Card Fraud Detection dataset before performing exploratory data analysis.

This notebook focuses on:

- Understanding dataset dimensions
- Identifying feature types
- Checking missing values
- Detecting duplicate records
- Understanding the target variable distribution
- Assessing overall data quality

The findings from this notebook will guide the preprocessing and exploratory data analysis phases.

## 2. Import Libraries

Import necessary libraries for data manipulation and basic display.

In [16]:
import pandas as pd
import numpy as np
from pathlib import Path

# Set display options for better readability
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:.4f}'.format)

## 3. Load Dataset

Load the raw dataset from the data directory.

In [17]:
DATA_PATH = Path(r"D:\Fraud Detection\data\raw\creditcard.csv")
df = pd.read_csv(DATA_PATH)
print("Dataset loaded successfully.")

Dataset loaded successfully.


## 4. Dataset Overview

Let's take a quick look at the first few rows to get a sense of the features and values.

In [18]:
df.head()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12,V13,V14,V15,V16,V17,V18,V19,V20,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
0,0.0000,-1.3598,-0.0728,2.5363,1.3782,-0.3383,0.4624,0.2396,0.0987,0.3638,0.0908,-0.5516,-0.6178,-0.9914,-0.3112,1.4682,-0.4704,0.2080,0.0258,0.4040,0.2514,-0.0183,0.2778,-0.1105,0.0669,0.1285,-0.1891,0.1336,-0.0211,149.6200,0
1,0.0000,1.1919,0.2662,0.1665,0.4482,0.0600,-0.0824,-0.0788,0.0851,-0.2554,-0.1670,1.6127,1.0652,0.4891,-0.1438,0.6356,0.4639,-0.1148,-0.1834,-0.1458,-0.0691,-0.2258,-0.6387,0.1013,-0.3398,0.1672,0.1259,-0.0090,0.0147,2.6900,0
2,1.0000,-1.3584,-1.3402,1.7732,0.3798,-0.5032,1.8005,0.7915,0.2477,-1.5147,0.2076,0.6245,0.0661,0.7173,-0.1659,2.3459,-2.8901,1.1100,-0.1214,-2.2619,0.5250,0.2480,0.7717,0.9094,-0.6893,-0.3276,-0.1391,-0.0554,-0.0598,378.6600,0
3,1.0000,-0.9663,-0.1852,1.7930,-0.8633,-0.0103,1.2472,0.2376,0.3774,-1.3870,-0.0550,-0.2265,0.1782,0.5078,-0.2879,-0.6314,-1.0596,-0.6841,1.9658,-1.2326,-0.2080,-0.1083,0.0053,-0.1903,-1.1756,0.6474,-0.2219,0.0627,0.0615,123.5000,0
4,2.0000,-1.1582,0.8777,1.5487,0.4030,-0.4072,0.0959,0.5929,-0.2705,0.8177,0.7531,-0.8228,0.5382,1.3459,-1.1197,0.1751,-0.4514,-0.2370,-0.0382,0.8035,0.4085,-0.0094,0.7983,-0.1375,0.1413,-0.2060,0.5023,0.2194,0.2152,69.9900,0


In [26]:
# Last Five Rows
df.tail()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12,V13,V14,V15,V16,V17,V18,V19,V20,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
284802,172786.0000,-11.8811,10.0718,-9.8348,-2.0667,-5.3645,-2.6068,-4.9182,7.3053,1.9144,4.3562,-1.5931,2.7119,-0.6893,4.6269,-0.9245,1.1076,1.9917,0.5106,-0.6829,1.4758,0.2135,0.1119,1.0145,-0.5093,1.4368,0.2500,0.9437,0.8237,0.7700,0
284803,172787.0000,-0.7328,-0.0551,2.0350,-0.7386,0.8682,1.0584,0.0243,0.2949,0.5848,-0.9759,-0.1502,0.9158,1.2148,-0.6751,1.1649,-0.7118,-0.0257,-1.2212,-1.5456,0.0596,0.2142,0.9244,0.0125,-1.0162,-0.6066,-0.3953,0.0685,-0.0535,24.7900,0
284804,172788.0000,1.9196,-0.3013,-3.2496,-0.5578,2.6305,3.0313,-0.2968,0.7084,0.4325,-0.4848,0.4116,0.0631,-0.1837,-0.5106,1.3293,0.1407,0.3135,0.3957,-0.5773,0.0014,0.2320,0.5782,-0.0375,0.6401,0.2657,-0.0874,0.0045,-0.0266,67.8800,0
284805,172788.0000,-0.2404,0.5305,0.7025,0.6898,-0.3780,0.6237,-0.6862,0.6791,0.3921,-0.3991,-1.9338,-0.9629,-1.0421,0.4496,1.9626,-0.6086,0.5099,1.1140,2.8978,0.1274,0.2652,0.8000,-0.1633,0.1232,-0.5692,0.5467,0.1088,0.1045,10.0000,0
284806,172792.0000,-0.5334,-0.1897,0.7033,-0.5063,-0.0125,-0.6496,1.5770,-0.4147,0.4862,-0.9154,-1.0405,-0.0315,-0.1881,-0.0843,0.0413,-0.3026,-0.6604,0.1674,-0.2561,0.3829,0.2611,0.6431,0.3768,0.0088,-0.4736,-0.8183,-0.0024,0.0136,217.0000,0


In [27]:
# Random Sample
df.sample(5, random_state=42)

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,V11,V12,V13,V14,V15,V16,V17,V18,V19,V20,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
43428,41505.0000,-16.5265,8.5850,-18.6499,9.5056,-13.7938,-2.8324,-16.7017,7.5173,-8.5071,-14.1102,5.2992,-10.8340,1.6711,-9.3739,0.3608,-9.8992,-19.2363,-8.3986,3.1017,-1.5149,1.1907,-1.1277,-2.3586,0.6735,-1.4137,-0.4628,-2.0186,-1.0428,364.1900,1
49906,44261.0000,0.3398,-2.7437,-0.1341,-1.3857,-1.4514,1.0159,-0.5244,0.2241,0.8997,-0.5650,-0.0877,0.9794,0.0769,-0.2179,-0.1368,-2.1429,0.1270,1.7527,0.4325,0.5060,-0.2134,-0.9425,-0.5268,-1.1570,0.3112,-0.7466,0.0410,0.1020,520.1200,0
29474,35484.0000,1.3996,-0.5907,0.1686,-1.0300,-0.5398,0.0404,-0.7126,0.0023,-0.9717,0.7568,0.5438,0.1125,1.0754,-0.2458,0.1805,1.7699,-0.5332,-0.5333,1.1922,0.2129,0.1024,0.1683,-0.1666,-0.8102,0.5051,-0.2323,0.0114,0.0046,31.0000,0
276481,167123.0000,-0.4321,1.6479,-1.6694,-0.3495,0.7858,-0.6306,0.2770,0.5860,-0.4847,-1.3766,-1.3283,0.2236,1.1326,-0.5509,0.6166,0.4980,0.5022,0.9813,0.1013,-0.2446,0.3589,0.8737,-0.1786,-0.0172,-0.2074,-0.1578,-0.2374,0.0019,1.5000,0
278846,168473.0000,2.0142,-0.1374,-1.0158,0.3273,-0.1822,-0.9566,0.0432,-0.1607,0.3632,0.2595,0.9422,0.8500,-0.6162,0.5926,-0.6038,0.0911,-0.4719,-0.3338,0.4047,-0.2553,-0.2386,-0.6164,0.3470,0.0616,-0.3602,0.1747,-0.0780,-0.0706,0.8900,0


## 5. Dataset Structure

Check the dimensions of the dataset and basic memory usage information.

In [19]:
print(f"Number of Rows: {df.shape[0]:,}")
print(f"Number of Columns: {df.shape[1]}")
print("\n--- Dataset Info ---")
df.info(memory_usage='deep')

Number of Rows: 284,807
Number of Columns: 31

--- Dataset Info ---
<class 'pandas.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  floa

In [28]:
# Column Names
df.columns.tolist()

['Time',
 'V1',
 'V2',
 'V3',
 'V4',
 'V5',
 'V6',
 'V7',
 'V8',
 'V9',
 'V10',
 'V11',
 'V12',
 'V13',
 'V14',
 'V15',
 'V16',
 'V17',
 'V18',
 'V19',
 'V20',
 'V21',
 'V22',
 'V23',
 'V24',
 'V25',
 'V26',
 'V27',
 'V28',
 'Amount',
 'Class']

## 6. Data Types

Ensure all columns have the expected data types (numerical for this dataset).

In [29]:
df.dtypes

Time      float64
V1        float64
V2        float64
V3        float64
V4        float64
V5        float64
V6        float64
V7        float64
V8        float64
V9        float64
V10       float64
V11       float64
V12       float64
V13       float64
V14       float64
V15       float64
V16       float64
V17       float64
V18       float64
V19       float64
V20       float64
V21       float64
V22       float64
V23       float64
V24       float64
V25       float64
V26       float64
V27       float64
V28       float64
Amount    float64
Class       int64
dtype: object

In [20]:
df.dtypes.value_counts()

float64    30
int64       1
Name: count, dtype: int64

## 7. Missing Value Analysis

Check if there are any missing (NaN/Null) values in the dataset that would require imputation.

In [21]:
missing_values = df.isnull().sum()
missing_cols = missing_values[missing_values > 0]

if missing_cols.empty:
    print("No missing values found in the dataset.")
else:
    print("Missing values found:\n", missing_cols)

No missing values found in the dataset.


In [31]:
# Missing Percentage
missing_percentage = (missing_values / len(df)) * 100

missing_percentage

Time     0.0000
V1       0.0000
V2       0.0000
V3       0.0000
V4       0.0000
V5       0.0000
V6       0.0000
V7       0.0000
V8       0.0000
V9       0.0000
V10      0.0000
V11      0.0000
V12      0.0000
V13      0.0000
V14      0.0000
V15      0.0000
V16      0.0000
V17      0.0000
V18      0.0000
V19      0.0000
V20      0.0000
V21      0.0000
V22      0.0000
V23      0.0000
V24      0.0000
V25      0.0000
V26      0.0000
V27      0.0000
V28      0.0000
Amount   0.0000
Class    0.0000
dtype: float64

## 8. Duplicate Analysis

Identify if there are any perfectly identical rows in the dataset. While some duplicates might be legitimate in transactional data, it's important to quantify them.

In [22]:
duplicate_count = df.duplicated().sum()
duplicate_pct = (duplicate_count / len(df)) * 100

print(f"Number of duplicate rows: {duplicate_count:,}")
print(f"Percentage of duplicates: {duplicate_pct:.4f}%")

Number of duplicate rows: 1,081
Percentage of duplicates: 0.3796%


## 9. Target Variable Analysis

Examine the distribution of the `Class` column to quantify the class imbalance.

In [23]:
class_counts = df['Class'].value_counts()
class_pcts = df['Class'].value_counts(normalize=True) * 100

target_dist = pd.DataFrame({
    'Count': class_counts,
    'Percentage (%)': class_pcts
})

target_dist.index = ['Legitimate (0)', 'Fraudulent (1)']
target_dist

,Count,Percentage (%)
Legitimate (0),284315,99.8273
Fraudulent (1),492,0.1727


## 10. Descriptive Statistics

Generate summary statistics for numerical columns to understand their scale and distribution. Note that `V1`–`V28` are PCA components, so they will be roughly centered around 0.

In [32]:
# Numerical Summary
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Time,284807.0000,94813.8596,47488.1460,0.0000,54201.5000,84692.0000,139320.5000,172792.0000
V1,284807.0000,0.0000,1.9587,-56.4075,-0.9203,0.0180,1.3156,2.4549
V2,284807.0000,0.0000,1.6513,-72.7157,-0.5985,0.0655,0.8037,22.0577
V3,284807.0000,0.0150,8.0739,-48.3256,-0.8904,0.1798,1.0272,4232.0000
V4,284807.0000,0.0000,1.4159,-5.6832,-0.8486,-0.0198,0.7434,16.8753
V5,284807.0000,0.0000,1.3802,-113.7433,-0.6916,-0.0543,0.6119,34.8017
V6,284807.0000,-0.0000,1.3322,-26.1605,-0.7683,-0.2742,0.3985,73.3016
V7,284807.0000,0.0000,1.2371,-43.5572,-0.5540,0.0401,0.5704,120.5895
V8,284807.0000,0.0000,1.1944,-73.2167,-0.2086,0.0224,0.3273,20.0072
V9,284807.0000,0.0043,2.5380,-13.4341,-0.6431,-0.0513,0.5971,1221.0000


In [24]:
# Summarize Time, Amount, and a few PCA components to avoid cluttering the output
df[['Time', 'Amount', 'V1', 'V2', 'V3', 'V28', 'Class']].describe()

,Time,Amount,V1,V2,V3,V28,Class
count,284807.0000,284807.0000,284807.0000,284807.0000,284807.0000,284807.0000,284807.0000
mean,94813.8596,88.3496,0.0000,0.0000,0.0150,-0.0000,0.0017
std,47488.1460,250.1201,1.9587,1.6513,8.0739,0.3301,0.0415
min,0.0000,0.0000,-56.4075,-72.7157,-48.3256,-15.4301,0.0000
25%,54201.5000,5.6000,-0.9203,-0.5985,-0.8904,-0.0530,0.0000
50%,84692.0000,22.0000,0.0180,0.0655,0.1798,0.0112,0.0000
75%,139320.5000,77.1650,1.3156,0.8037,1.0272,0.0783,0.0000
max,172792.0000,25691.1600,2.4549,22.0577,4232.0000,33.8478,1.0000


## 11. Initial Data Quality Assessment

Based on the checks above, we assess the overall quality of the dataset:
- **Missing Data:** Clean / Needs Imputation
- **Data Types:** Correct / Needs Conversion
- **Duplicates:** Present / Absent
- **Anomalies:** Scale differences (Time/Amount vs PCA components)

## 12. Key Observations

1. **Imbalance:** The dataset is heavily imbalanced, with frauds accounting for only ~0.17%.
2. **Scaling:** The PCA components (`V1`–`V28`) are already scaled/transformed. However, `Time` and `Amount` are in their original scales and will need normalization/standardization before feeding into distance-based models or neural networks.
3. **Completeness:** There are no missing values in this dataset.
4. **Duplicates:** There is a small percentage of duplicate rows. We will need to decide whether to drop them in the preprocessing stage.

## 13. Next Steps

Proceed to **`03_EDA.ipynb`** for:
- Visualizing feature distributions for both classes.
- Correlation analysis to see relationships between features.
- Bivariate analysis focusing on `Time` and `Amount` vs `Class`.